paddle/fluid/framework/details/inplace_op_pass.cc

```diff
@@ -0,0 +1,362 @@
+    "sigmoid",
+    "exp",
+    "relu",
+    "tanh",
+    "sqrt",
+    "ceil",
+    "floor",
+    "reciprocal",
+    "relu6",
+    "soft_relu",
+    "hard_sigmoid",
+    "batch_norm",
+    "batch_norm_grad",
+    "sum",
+    "sum_grad",
+    "scale",
+    "reshape",
+    "elementwise_add",
+    "elementwise_add_grad",
+};
+// clang-format on
+
+namespace paddle {
+namespace framework {
+namespace details {
+
+static inline ir::Node* GetNextInplacedOpOutput(ir::Node* var) {
+  // if next op is inplaced, then return the output var
+  // otherwise return nullptr
+  PADDLE_ENFORCE(var && var->IsVar() && !var->IsCtrlVar());
+  ir::Node* inplaced_var = nullptr;
+  // only has one output op can be inplaced
+  if (var->outputs.size() == 1 && var->outputs[0]->IsOp()) {
+    auto* op = var->outputs[0];
+    for (auto* out_var : op->outputs) {
+      if (!out_var->IsVar() || out_var->IsCtrlVar() ||
+          out_var->Var() == nullptr)
+        continue;
+      if (out_var->Name() == var->Name()) {
+        inplaced_var = out_var;
+        break;
+      }
+    }
+  }
+  return inplaced_var;
+}
+
+static inline ir::Node* GetPrevInplacedOpInput(ir::Node* var) {
+  PADDLE_ENFORCE(var && var->IsVar() && !var->IsCtrlVar());
+  ir::Node* inplaced_var = nullptr;
+  if (var->inputs.size() == 1 && var->inputs[0]->IsOp()) {
+    auto* op = var->inputs[0];
+    for (auto* in_var : op->inputs) {
+      if (!in_var->IsVar() || in_var->IsCtrlVar() || in_var->Var() == nullptr)
+        continue;
+      if (in_var->Name() == var->Name()) {
+        inplaced_var = in_var;
+        break;
+      }
+    }
+  }
+  return inplaced_var;
+}
+
+template <typename Container>
+static inline bool ConnectByCtrlVar(const Container& group1,
+                                    const Container& group2) {
+  bool connected = false;
+  std::unordered_set<ir::Node*> outputs;
+  for (auto* op : group1) {
+    for (auto* var : op->outputs) {
+      if (var->IsCtrlVar()) outputs.emplace(var);
+    }
+  }
+  for (auto* op : group2) {
+    for (auto* var : op->inputs) {
+      if (outputs.count(var)) connected = true;
+    }
+  }
+  return connected;
+}
+
+InplacePass::InplacePass() : Pass() {
+  if (FLAGS_enable_inplace_whitelist) {
+    for (auto& s : kInplacedOpWhiteList) {
+      whitelist_.emplace(s);
+    }
+  }
+}
+
+void InplacePass::InitSSAGraphNodes() const {
+  std::unordered_map<std::string, std::unordered_set<ir::Node*>> all_vars;
+  for (auto* op : view_.AllOps()) {
+    for (auto* node : op->inputs) {
+      if (!node->IsVar() || node->IsCtrlVar()) continue;
+      if (all_vars[node->Name()].count(node) == 0) {
+        all_vars[node->Name()].emplace(node);
+        var_nodes_[node->Name()].emplace_back(node);
+      }
+    }
+    for (auto* node : op->outputs) {
+      if (!node->IsVar() || node->IsCtrlVar()) continue;
+      if (all_vars[node->Name()].count(node) == 0) {
+        all_vars[node->Name()].emplace(node);
+        var_nodes_[node->Name()].emplace_back(node);
```

Comment:
It works same way as the https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/ir/graph.cc#L48
because there are some paramters/optional (dispensable), so we also need to track the inputs. Otherwise the resolvedataharzards wil failed.